### PR TITLE
Extract and map

### PR DESCRIPTION
**Pull request by mestachs**
really life example (assuming you the credentials to access the server)

```python
from io import StringIO
import numpy as np
import pandas as pd
import code
import json

from blsqpy.descriptor import Descriptor
from blsqpy.periods import Periods

from blsqpy.s3export_hook import S3ExportsHook
from blsqpy.postgres_hook import PostgresHook
from blsqpy.dhis2 import Dhis2

import blsqpy.extract as extract
import blsqpy.mapping as mapping

r = {
    "activities": {
        "pills": {
            "name": "Pilule",
            "states": {
                "active": {
                    "sources": {
                        "moh": {
                            "uids": ["s4CxsmoqdRj.oFtySBIkit2"]
                        }
                    }
                },
                "new_structure": {
                    "sources": {
                        "moh": {
                            "uids": ["s4CxsmoqdRj.pQqm0IZ6AJ3"]
                        }
                    }
                },
                "new_program": {
                    "sources": {
                        "moh": {
                            "uids": ["s4CxsmoqdRj.Vkfh73dsruW"]
                        }
                    }
                },
                "delivered": {
                    "sources": {
                        "moh": {
                            "uids": [
                                "fSD1ZZo4hTs.oFtySBIkit2",
                                "p4GY25ry19H.pQqm0IZ6AJ3",
                                "sur5ZhRRQYp.HllvX50cXC0"
                            ],
                            "comments": [
                                "Pilules : Microgynon distribuée",
                                "Pilules : Autres distribuée",
                                "Pilules : Microlut distribuée"
                            ]
                        }
                    }
                }
            }
        }
    }
}


json_string = json.dumps(r)
config = Descriptor.load_string(json_string)
activity = config.activities.pills


# start extract -------------------------------------------

dhis = Dhis2(PostgresHook('dhis2_sn_cartesanitaire'))
df = dhis.get_data(extract.to_data_elements(activity))
df = extract.rotate_de_coc_as_columns(df)

# start mapping
activity = config.activities.pills
df = mapping.map_from_activity(df, activity, "pills")

print(df)

print(df.query("orgunit == 'skBfuNMAMq1'"))

```

### CODE REVIEW

File: specs/config/sample.json [hunk 1]
```diff
@@ -0,0 +1,47 @@
+{
+  "default_category_option_combos": "HllvX50cXC0",
```
Comment (grlurton) [inline specs/config/sample.json]: Just as reminder, as mentioned on Slack, we probably want to have this inside specific data elements defs

File: specs/extract_spec.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+from blsqpy.descriptor import Descriptor
+import blsqpy.extract as extract
+from pandas.testing import assert_frame_equal
+
+with description('extract') as self:
+
+    with it('extract data elements from config'):
+        config = Descriptor.load("./specs/config/sample")
+
+        data_elements = extract.to_data_elements(config.activities.pills)
```
Comment (grlurton) [inline specs/extract_spec.py]: Should we have a check that we don't have duplicate activities in config ?

Comment (mestachs) [reply specs/extract_spec.py]: as they are build like json "Dict" at least the "code" (pills) is unique

File: specs/config/sample.json (hunk 1 above)
Comment (mestachs) [reply specs/config/sample.json]: thanks, just removed it.

File: blsqpy/extract.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+def to_data_elements(activity):
+    data_elements = []
+    for _state_code, state in Descriptor.as_items(activity.states):
+        for _source_code, source in Descriptor.as_items(state.sources):
+            data_elements.extend(source.uids)
+    return data_elements
+
+
+def rotate_de_coc_as_columns(df):
+    df = df.rename(index=str, columns={
```
Comment (grlurton) [inline blsqpy/extract.py]: Let's standardize our naming earlier (on df or config) so we stay consistent in naming accross the pipe

File: blsqpy/extract.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+        "uidorgunit": "orgunit",
+        "catcomboid": "coc",
+        "dataelementid": "de"
+    })
+
+    df["period"] = np.where(df.monthly, df.monthly, df.quarterly)
+    df = df.drop(["uidlevel3", "uidlevel2", "enddate",
```
Comment (grlurton) [inline blsqpy/extract.py]: We want to keep "uidlevel3", "uidlevel2" in at that point as we use them for aggregation.

Also, depending on extracts / countries, the relevant grouping will be a different level. would need to paramatrize that accross extraction and mapping.

File: blsqpy/mapping.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+            if len(source.uids) > 1:
+                state_expressions = []
+                for idx in range(len(source.uids)):
+                    state_expressions.append(
+                        prefix+state_code+"_"+str(idx+1)+suffix)
+                expressions[prefix+state_code+suffix] = state_expressions
+    return expressions
+
+
+def to_mappings(activity, activity_code):
```
Comment (grlurton) [inline blsqpy/mapping.py]: I don't fully understand at what point we need those, but we should probably refactor into one function ?

File: blsqpy/extract.py (hunk 4 above)
Comment (mestachs) [reply blsqpy/extract.py]: I think ou will be part of another data frame that you will join later (at aggregation)

File: blsqpy/mapping.py (hunk 5 above)
Comment (mestachs) [reply blsqpy/mapping.py]: `to_mappings `
   help translating the previous dataframe format which is the form ou, period, de2.category2, de2.category2,... value
   see the rotated.csv  in the PR
   in more "business" but less brittle format than "just de name"
   ou, period,  activity_state1_source, activity_state2_source, activity_state3_source
   see mapped.csv

`to_expressions`
a last step if you "provided multiple de.category1 in the state.source" is to sum
   eg in the spec the delivered columns so pills_delivered_moh =   ('fSD1ZZo4hTs.oFtySBIkit2', 'pills_delivered_1_moh') +  ('p4GY25ry19H.pQqm0IZ6AJ3', 'pills_delivered_2_moh') +              ('sur5ZhRRQYp.HllvX50cXC0', 'pills_delivered_3_moh'),

   think "dhis2 indicator that just sums multiple category combo"

Comment (mestachs) [reply blsqpy/mapping.py]: both are used "below" in the map_from_activity(df, activity, activity_code)

**Pull request merged**